### PR TITLE
Adds management command to generate enrollment codes for legacy learners

### LINKED ISSUE #961: generate enrollment codes for paid but not enrolled legacy users
As a legacy MicroMasters user who:

1. Is **not** enrolled in a current MITx Online course (3T2022 or 1T2023)
2. has paid
3. has not used all my exam attempts , 

I'd like to contact CS to request an enrollment code (100% off coupon code/discount) for future use. 

As a CS teammember, I'd like to be able to distribute enrollment codes for anyone who has paid, so they can enroll in a verified track and take the proctored exam when they like. 

#### Acceptance Criteria:

- [ ] generate a list of users (emails) and course runs of users who have paid, but aren't enrolled in a current run 
- [ ] write a management command or similar to create 100% off discounts on mitxoline for these learners
- [ ] since the codes will be distributed only on request, we could generate this once and not be concerned with updates. Any user paying between now and the launch of ecommerce is expected to enroll. 
- [ ] Since these users have paid on micromasters, and AR/Finance has already accounted for their payments. The discount codes are primarily a convenience for the user, but we should consider a naming scheme that makes it easy for AR to recognize and ignore them. 

#### Related issues 

- [ ] test importing of grades into open edX  
- [x] #960

### PR DESCRIPTION
**Pull request by jkachel**
#### Pre-Flight checklist

- [X] Testing
  - [X] Code is tested
  - [X] Changes have been manually tested

#### What are the relevant tickets?

Fixes #961 

#### What's this PR do?

Adds a management command, `generate_legacy_enrollment_codes`, that will generate a 100% off discount code for learners that have paid for a courserun but aren't enrolled for it. (See #960.) The discount codes are:
* 100% off
* One-time use
* Associated with the specified course run
* Set as a user discount (so, barring other discounts, they will autoapply when the learner attempts to purchase the course)
* Formatted `MM-legacy-<username>-<readable_id>`
The generated codes and the email addresses for the learners they apply to are written to a CSV file.

This also adds a management command, `find_paid_unenrolled_learners`, which will look in the _MITxOnline_ database for users that have purchased a courserun but aren't enrolled in it at all. This outputs in a format that can be ingested by the `generate_legacy_enrollment_codes` command.

#### How should this be manually tested?

The `generate_legacy_enrollment_codes` checks for the existence of the learner and the specified courserun but nothing else, so it can be tested with any learner that exists in the system. 

1. Generate an input CSV file. The file should be in the format `learner email,readable id`. 
2. Run the `generate_legacy_enrollment_codes` command: `./manage.py generate_legacy_enrollment_codes <learner list file> <code output file>`
3. Check that the discount codes are created and set up correctly in Django Admin.

Once the codes are generated, a learner that doesn't have any other automatic discounts (like, for instance, financial assistance) should have the newly generated discount applied when they add the course to their cart automatically. If the learner does have another discount, the code should work to apply it manually. (Financial assistance discounts take precedence over other discounts other than ones that are manually applied.)

### CODE REVIEW

**Comment by rachellougee** [comment]
The code looks good to me. For `find_paid_unenrolled_learners` command, it works as expected. 
But I got errors when running `generate_legacy_enrollment_codes`, It looks like some discount code is too long and there is a limit of 50 characters. After I remove a few characters in the code, it then ran fine. so we might want to expand discount_code?

**Review by rachellougee — APPROVED** [review]

**Comment by jkachel** [comment]
@pdpinch thoughts on the generated enrollment code format? I erred on the side of more or less human readable but we could do something less so (like, say, `MM-legacy-<user_id>-<courserun_id>` (or something completely random or something else). Or, I can keep it as is and up the length limit on the codes; I don't recall there being a real reason for the code length to be 50 other than maybe that's how it was in MicroMasters.

**Comment by annagav** [comment]
@jkachel I don't think that we can provide course run key for the input file for this task, it's just a course id.

**Comment by jkachel** [comment]
Ok - I updated in https://github.com/mitodl/mitxonline/pull/1039/commits/08edfc5c5596562efda14366b10d4cd0933a2051 to look for a matching course object as well. It'll use the `first_unexpired_run` to determine what courserun to use in that case (since products can't be for a course, and as such neither can discounts).

**Review by rachellougee — APPROVED** [review]

**Comment by rachellougee** [comment]
The updated code looks good to me. 
I tested a couple of course IDs, and it was able to generate discount code for unexpired run for course ID. 
Just need to figure out the format of the coupon code in case someone's user name is too long

File: ecommerce/management/commands/generate_legacy_enrollment_codes.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                        self.style.ERROR(
+                            f"Can't find account for user {row[0]}, skipping row"
+                        )
+                    )
+                    continue
+
+                courserun = CourseRun.objects.filter(courseware_id=row[1]).first()
+
+                if courserun is None:
+                    course = Course.objects.filter(readable_id=row[1]).first()
```
Comment (annagav) [inline ecommerce/management/commands/generate_legacy_enrollment_codes.py:61]: I think here you might need to map the `course_id` from MM into `readable_id` in MITxOnline through this mapping model CourseId from micromasters_import/models.py.

**Review by annagav — CHANGES_REQUESTED** [review]

File: ecommerce/management/commands/generate_legacy_enrollment_codes.py (hunk 1 above)
Comment (rachellougee) [reply ecommerce/management/commands/generate_legacy_enrollment_codes.py:61]: CourseId in micromasters_import/models.py is the auto-incremented primary key of the course model, not the string field https://rc.mitxonline.mit.edu/admin/micromasters_import/courseid/7/change/. Are we exporting the ID 7 to csv from https://micromasters-rc.odl.mit.edu/admin/courses/course/7/change/?

**Comment by pdpinch** [comment]
We should use something less transparent, but still filterable, like: MM22-<uuid>

**Comment by jkachel** [comment]
https://github.com/mitodl/mitxonline/pull/1039/commits/ca4f92bdc673c023e6cdedc3609f5d9f003c15d0 updates the code format and adds in a check for the numeric ID mapping if the supplied course ID is an integer.

**Review by annagav — APPROVED** [review]
👍

**Pull request merged**